_posts/2021-01-18-smtchecker-and-synthesis-of-external-functions.md

```diff
@@ -0,0 +1,420 @@
+a formal verification module built-in the Solidity compiler, received a lot of
+improvements of various types in 2020. Many improvements are related to
+Solidity's language features, such as structs, ABI, hash and math functions,
+and other important features that were unsupported.
+When running the tool, the user now has more control and is able to set a
+timeout per query and choose engines separately between
+[BMC and CHC](https://docs.soliditylang.org/en/v0.8.0/security-considerations.html#model-checking-engines).
+The CHC engine encodes Solidity code as [Constrained Horn Clauses](https://en.wikipedia.org/wiki/Horn_clause)
+and is often able to
+[handle loops and contract invariants](https://medium.com/@leonardoalt/smtchecker-toward-completeness-1a99c02e0133).
+The internals of the CHC engine will be detailed in an upcoming series of
+posts.
+CHC has been extended to analyze external calls to unknown code and report
+counterexamples including transaction traces, which are the focus of this post.
+
+## Counterexamples and Transaction Traces
+
+The CHC engine has been able to tell whether a property holds or not for a
+while, but counterexample reporting was only added recently.  The
+counterexample attempts to give concrete values for state, input and output
+variables.  It works well in general, but might fail sometimes for non-value
+types due to their complexity.  The user also receives a full transaction trace
+that leads to the property being violated.
+
+Consider the Solidity sample below:
+```solidity
+//SPDX-License-Identifier: GPL-v3
+pragma solidity ^0.8.0;
+pragma experimental SMTChecker;
+
+contract Test {
+	function max(uint[] memory _a) public pure returns (uint) {
+		require(_a.length > 5);
+		uint m = 0;
+		for (uint i = 0; i < _a.length; ++i)
+			if (_a[i] > m)
+				m = _a[i];
+
+		for (uint i = 0; i < _a.length; ++i)
+			assert(m > _a[i]);
+
+		return m;
+	}
+}
+```
+
+The function takes an array of length at least 6, and computes the maximum
+element of the array.  After that, it asserts that the previously found `max
+element is greater than all the elements of the array`, which is what the
+SMTChecker will try to prove.  The property is almost true:
+
+```bash
+$ solc max.sol --model-checker-engine chc
+
+Warning: CHC: Assertion violation happens here.
+Counterexample:
+
+_a = [0, 0, 0, 0, 0, 0]
+
+Transaction trace:
+Test.constructor()
+Test.max([0, 0, 0, 0, 0, 0])
+  --> max.sol:14:4:
+   |
+14 | 			assert(m > _a[i]);
+   | 			^^^^^^^^^^^^^^^^^
+```
+
+The SMTChecker gives us a pretty simple counterexample, an array full of
+zeroes. Our property missed the fact that the max element is also in the array,
+and therefore cannot be greater than itself. The correct property should be
+`assert(m >= _a[i]);`. After fixing the property, we're able to prove it, shown
+by the absence of a warning from the tool.
+Note that in the same run the SMTChecker also proved that none of the `++i` can
```

Comment:
hmm, isn't this given though since this is solc 0.8.0?

Reply:
Not exactly. 0.8 makes sure that every path that overflows reverts. The SMTChecker would still report overflows in that case. Here it proves that there's no path that would make i++ overflow